persistent/Database/Persist/Sql/TH.hs

```diff
@@ -0,0 +1,119 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE TemplateHaskellQuotes #-}
+
+module Database.Persist.Sql.TH
+  ( fromN
+  , generateRawSqlHelper
+  , generateRawSqlInstances
+  , toN
+  )
+where
+
+import Control.Monad
+import Language.Haskell.TH
+
+#if MIN_VERSION_template_haskell(2,16,0)
+prepTupE :: a -> Maybe a
+prepTupE = Just
+#else
+prepTupE :: a -> a
+prepTupE = id
+#endif
+
+minTupleSize :: Int
+minTupleSize = 3 -- instance for 2 is already written
+
+maxTupleSize :: Int
+maxTupleSize = 62 -- GHC limitation
+
+pairedList :: [a] -> [[a]]
+pairedList [] = []
+pairedList [x] = [[x]]
+pairedList (x:y:rest) = [x, y] : pairedList rest
```

Comment:
I feel like this would be better as a new data structure that captures the invariant we want.

```suggestion
data PairedList a 
  = Pair a a (PairedList a)
  | End (Maybe a)

pairedList :: [a] -> PairedList a
pairedList [] = End Nothing
pairedList [x] = End (Just x)
pairedList (x:y:rest) = Pair x y $ pairedList rest
```

Reply:
`PairedList` can be factored out to `([(a, a)], Maybe a)`, but a tuple of list of tuples sounds annoying to work with.